### PR TITLE
View optimization and installed crispy forms, added UI and registration module

### PR DESCRIPTION
**Pull request by annalitvin**
Added mixins to valid create forms.

### CODE REVIEW

File: src/teacher/views.py [hunk 1]
```diff
@@ -1,113 +1,127 @@
...
-            return HttpResponse(f'Data added fail! {ex}', status=500)
-        return redirect('teacher:data_success')
-    return HttpResponse("Error. The value must be numeric and greater than zero.", status=400)
+            id_teacher = int(id_teacher)
+            if id_teacher == 0:
+                return HttpResponse("'id' must be greatest when zero")
+
+        self.object = self.get_object()
+        context = self.get_context_data()
+        return self.render_to_response(context)
```
Comment (dbradul) [inline src/teacher/views.py:48]: it's better to call parent method:
```
        self.object = self.get_object()
        context = self.get_context_data()
        return self.render_to_response(context)
```
-->

```
        return super().get(request, *args, **kwargs)
```

File: src/teacher/views.py [hunk 2]
```diff
@@ -1,113 +1,127 @@
...
-        if not id_teacher.isdigit():
-            return HttpResponse("'id' must be numeric and greatest when zero")
+    def get_object(self, queryset=None):
 
-        id_teacher = int(id_teacher)
-        if id_teacher == 0:
-            return HttpResponse("'id' must be greatest when zero")
+        request = self.request
 
+        id_teacher = request.GET.get('id')
```
Comment (dbradul) [inline src/teacher/views.py:23]: It's better to use more high-level source of data as request params are already parsed by this moment

```
id_teacher = request.GET.get('id')
```
-->
```
id_teacher = self.kwargs.get(self.pk_url_kwarg, None)
```

File: src/teacher/views.py [hunk 3]
```diff
@@ -1,113 +1,127 @@
...
-        return render(request, "teacher/index.html", context={"teacher": teacher})
-    return HttpResponse("Введите параметр id: ?id=n")
+            group_obj = self.model.objects.get(pk=id_teacher)
+        except self.model.DoesNotExist:
+            raise Http404("Group does not exist! "
+                          "Generate teachers in database using url: generate?teach_number=1" +
+                          "where teach_number is number of teachers. Default: 100 teachers")
+        return group_obj
 
+    def get(self, request, *args, **kwargs):
```
Comment (dbradul) [inline src/teacher/views.py:32]: Typically, `DetailView` is registered with pk or slug based url, like `'/teacher/<pk:int>'`.

File: src/group/mixins.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+
+class GroupValidFormMixin:
+
+    def form_valid(self, form):
```
[the remark this answers is not in the document]
Comment (annalitvin) [reply src/group/mixins.py:4]: Thanks! Fixed!

File: src/teacher/views.py (hunk 1 above)
Comment (annalitvin) [reply src/teacher/views.py:48]: Fixed!

File: src/teacher/forms.py [hunk 5]
```diff
@@ -10,7 +12,17 @@ class Meta:
...
+    def clean(self):
+        cleaned_data = super().clean()
+        phone_number = cleaned_data.get('phone_number').strip()
+        email = cleaned_data.get('email').strip()
+
+        is_teacher_exists = Teacher.objects.filter(Q(phone_number=phone_number) |
+                                                   Q(email=email)).exists()
+        if is_teacher_exists:
+            error_massage = "Teacher not added. Teacher with such phone_number or email is exists! Try again:"
+            raise ValidationError(error_massage)
```
Comment (dbradul) [inline src/teacher/forms.py:25]: Consider returning `cleaned_date` here. For example, you might want to derive from this form and add some additional validation in derived class. Example: `FlatpageForm` from `django.auth.contrib.flatpages`

File: src/student/forms.py [hunk 6]
```diff
@@ -24,7 +23,33 @@ def clean(self):
...
+
+    def __init__(self, *args, **kwargs):
+        self.pk = kwargs.pop('instance', None)
+        super().__init__(*args, **kwargs)
+
+    def clean_email(self):
+        cleaned_data = super().clean()
+        email = cleaned_data.get('email').strip()
+
+        is_email_exists = Student.objects.filter(email=email).exclude(pk=self.pk).exists()
```
Comment (dbradul) [inline src/student/forms.py:39]: An instance that is currently being updated is available as `sefl.instance`. So, you can get access to its id without extracting from params in constructor.

**Pull request merged**